<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">
## Pandas Concatenation Lab

_Authors: Riley Dallas (ATX), Dave Yerrington (SF)_

Here's how you import a dataset from Scikit Learn:

In [1]:
from sklearn.datasets import load_iris
import pandas as pd

Scikit Learn datasets are actually functions that return an object containing the data we need. 

#### In the cell below, call the `load_iris()` function and set the result to a variable called `data`.

In [2]:
data = load_iris()

In machine learning, the column we're trying to predict is usually called the **target** (or **label**). 

#### To see the targets for our dataset, call `data['target']` in the cell below.

> Also, you can use `data.keys()` to see a list of elements available inside the `data` object.
>```python
> dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])
>```

In [3]:
data['target']

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In the case of the iris dataset, the target is the particular species of iris flower. Because machine learning requires our features and target to be numbers, the species are encoded as 0, 1, 2. These indices correspond to the labeled species on `data.target_names`

#### Call `data['target_names']` to see the actual names of the three species

In [4]:
data['target_names']

array(['setosa', 'versicolor', 'virginica'],
      dtype='<U10')

#### Get the length of `data['target']` to see how many flowers are in this dataset.

In [5]:
data['target'].shape

(150,)

#### Use `data['target']` to create your first pandas DataFrame, which is just a single column (`"species"`).

In [13]:
species = pd.DataFrame(data['target'], columns=["species"])
species.head()

species
0        0
1        0
2        0
3        0
4        0

The features for this dataset are on `data['data']`. There are 150 rows, one for each flower and 4 columns, one for each of the following features:
1. sepal length (cm)
2. sepal width (cm)
3. petal length (cm)
4. petal width (cm)

#### Output `data['data']` in the cell below to see the features for this dataset

In [14]:
data['data']

array([[ 5.1,  3.5,  1.4,  0.2],
       [ 4.9,  3. ,  1.4,  0.2],
       [ 4.7,  3.2,  1.3,  0.2],
       [ 4.6,  3.1,  1.5,  0.2],
       [ 5. ,  3.6,  1.4,  0.2],
       [ 5.4,  3.9,  1.7,  0.4],
       [ 4.6,  3.4,  1.4,  0.3],
       [ 5. ,  3.4,  1.5,  0.2],
       [ 4.4,  2.9,  1.4,  0.2],
       [ 4.9,  3.1,  1.5,  0.1],
       [ 5.4,  3.7,  1.5,  0.2],
       [ 4.8,  3.4,  1.6,  0.2],
       [ 4.8,  3. ,  1.4,  0.1],
       [ 4.3,  3. ,  1.1,  0.1],
       [ 5.8,  4. ,  1.2,  0.2],
       [ 5.7,  4.4,  1.5,  0.4],
       [ 5.4,  3.9,  1.3,  0.4],
       [ 5.1,  3.5,  1.4,  0.3],
       [ 5.7,  3.8,  1.7,  0.3],
       [ 5.1,  3.8,  1.5,  0.3],
       [ 5.4,  3.4,  1.7,  0.2],
       [ 5.1,  3.7,  1.5,  0.4],
       [ 4.6,  3.6,  1. ,  0.2],
       [ 5.1,  3.3,  1.7,  0.5],
       [ 4.8,  3.4,  1.9,  0.2],
       [ 5. ,  3. ,  1.6,  0.2],
       [ 5. ,  3.4,  1.6,  0.4],
       [ 5.2,  3.5,  1.5,  0.2],
       [ 5.2,  3.4,  1.4,  0.2],
       [ 4.7,  3.2,  1.6,  0.2],
       [ 4

#### Create a `features` DataFrame in pandas, using `data['data']` and `data['feature_names']` 
> You can use `pd.DataFrame()` to create a DataFrame, passing the `columns` parameter using `data['feature_names']`

In [16]:
features = pd.DataFrame(data['data'], columns=data['feature_names'])
features.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

#### Use `pd.concat` to combine the two dataframes.
> The concat method essentially squishes two dataframes together, along either Axis 0 or Axis 1
> Axis 0 here refers to the rows. If we concat along axis 0, we would be stacking one DataFrame on top of the other. That is not what we want to accomplish in this case. Instead, we want to put the two dataframes side-by-side. In order to do that, we concat the columns by using Axis 1.
> Review / research the documentation on the useage of this function.  

In [19]:
df = pd.concat([features, species], axis=1)
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   species  
0        0  
1        0  
2        0  
3        0  
4        0

#### Compare concat with `DataFrame.join` by joining the two original DataFrames instead.

> In this case, we do not have to specify an axis. Join here is explicitly for joining two or more dataframes on columns. You can even pass a list of DataFrames to join if you have more than one (they must all have the same column to join on). Refer to the documentation [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html)

In [52]:
df = features.join(species)
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   species  
0        0  
1        0  
2        0  
3        0  
4        0

#### Compare concat with `DataFrame.merge` by merging the two original DataFrames instead.

> Like join, merge does not function across rows. Additionally, you will not that without a common column to merge on, you must be explicit about telling pandas to merge on the right and left indices. You can do this with `left_index=True` and `right_index=True`. Because we are not merging on a column, we do not pass `on` or `how` parameters, but both of those parameters will become very important to you as you do more sophisticated merging. Refer to the documentation [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html)

In [53]:
df = features.merge(species, left_index=True, right_index=True)
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   species  
0        0  
1        0  
2        0  
3        0  
4        0

#### Create a list of all the numeric column names by dropping the species column, then calling the `columns` attribute followed by the `values` attribute. Save it to a variable named `num_cols`. 

In [54]:
num_cols = df.drop('species', axis=1).columns.values

#### Multiply the four numeric columns together and save them as a new column `interaction`. 

In [55]:
df['interaction'] = df[num_cols[0]] * df[num_cols[1]] * df[num_cols[2]] * df[num_cols[2]]
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   species  interaction  
0        0      34.9860  
1        0      28.8120  
2        0      25.4176  
3        0      32.0850  
4        0      35.2800

#### Use the `.loc` method to display all of the rows (`:`) and only the numeric columns

In [56]:
df.loc[:,num_cols].head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

#### Z-score scale the columns except for species by combining a drop with an apply, and then overwriting the original columns. 

In [61]:
df.loc[:,np.append(num_cols, "interaction")] = df.drop("species", axis=1).apply(lambda x: (x - x.mean())/x.std(ddof=0))
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0          -0.900681          1.032057          -1.341272         -1.312977   
1          -1.143017         -0.124958          -1.341272         -1.312977   
2          -1.385353          0.337848          -1.398138         -1.312977   
3          -1.506521          0.106445          -1.284407         -1.312977   
4          -1.021849          1.263460          -1.341272         -1.312977   

   species  interaction  
0        0    -1.040388  
1        0    -1.062493  
2        0    -1.074646  
3        0    -1.050775  
4        0    -1.039336

#### Create a string (object in pandas) encoded column from the species column for human readable labels. Write a dictionary  `species_name` by using a dictionary comprehension on the `data['target_names']` array from the original iris dataset. You might consider the `enumerate` function as an easy way to accomplish this. Then use the `map` pandas method on the `species` column with the `species_name` dict. Save the result to `df['labels']`

In [64]:
species_name = {i:val for i, val in enumerate(data['target_names'])}
df['labels'] = df['species'].map(species_name)
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0          -0.900681          1.032057          -1.341272         -1.312977   
1          -1.143017         -0.124958          -1.341272         -1.312977   
2          -1.385353          0.337848          -1.398138         -1.312977   
3          -1.506521          0.106445          -1.284407         -1.312977   
4          -1.021849          1.263460          -1.341272         -1.312977   

   species  interaction  labels  
0        0    -1.040388  setosa  
1        0    -1.062493  setosa  
2        0    -1.074646  setosa  
3        0    -1.050775  setosa  
4        0    -1.039336  setosa